base/dates/io-util.jl

```diff
@@ -0,0 +1,178 @@
+### Parsing utilities
+
+macro chk1(expr,label=:error)
+    quote
+        x = $(esc(expr))
+        if isnull(x[1])
+            @goto $label
+        else
+            get(x[1]),x[2]
+        end
+    end
+end
+
+
+@generated function _tryparse{T, N}(fmt::DateFormat{T, NTuple{N}}, str::AbstractString)
+    quote
+        R = Nullable{NTuple{7,Int}}
+        t = fmt.tokens
+        l = fmt.locale
+        len = endof(str)
+
+        state = start(str)
+        err_idx = 1
+        Base.@nexprs $N i->val_i = 0
+        Base.@nexprs $N i->(begin
+            state > len && @goto done
+            (val_i, state) = @chk1 tryparsenext(t[i], str, state, len, l)
+            err_idx += 1
+        end)
+        state <= len && @goto error
+
+        @label done
+        parts = Base.@ntuple $N val
+        return R(reorder_args(parts, fmt.field_order, fmt.field_defaults, err_idx)::NTuple{7,Int})
+
+        @label error
+        return R((err_idx,state,0,0,0,0,0), true)
+    end
+end
+
+function tryfailparse{T}(dt, df::DateFormat{T})
+    maybedt = _tryparse(df, dt)
+    if isnull(maybedt)
+        err_data = maybedt.value # Unsafe! but _tryparse loads error data here
+        err_idx = err_data[1]
+        state = err_data[2]
+        # TODO: pretty print this output
+        if err_idx > length(df.tokens)
+            throw(ArgumentError("Found extra tokens at the end of date time string"))
+        else
+            throw(ArgumentError("Unable to parse date string. At token $(df.tokens[err_idx]) at char $(state)"))
+        end
+    else
+        _create_timeobj(maybedt.value, T)
+    end
+end
+
+_create_timeobj(tup, T::Type{DateTime}) = T(tup)
```

Comment:
I fixed changed this in a later commit. this review business is a bit confusing. Thanks for reviewing!